src/main/telemetry/ghst.c

```diff
@@ -0,0 +1,226 @@
+#include <stdint.h>
+#include <string.h>
+
+#include "platform.h"
+
+#ifdef USE_TELEMETRY_GHST
+
+#include "build/atomic.h"
+#include "build/build_config.h"
+#include "build/version.h"
+
+#include "config/feature.h"
+#include "pg/pg.h"
+#include "pg/pg_ids.h"
+
+#include "common/crc.h"
+#include "common/maths.h"
+#include "common/printf.h"
+#include "common/streambuf.h"
+#include "common/utils.h"
+
+#include "cms/cms.h"
+
+#include "drivers/nvic.h"
+
+#include "config/config.h"
+#include "fc/rc_modes.h"
+#include "fc/runtime_config.h"
+
+#include "flight/imu.h"
+#include "flight/position.h"
+
+#include "io/gps.h"
+#include "io/serial.h"
+
+#include "rx/ghst.h"
+#include "rx/ghst_protocol.h"
+
+#include "sensors/battery.h"
+#include "sensors/sensors.h"
+
+#include "telemetry/telemetry.h"
+#include "telemetry/msp_shared.h"
+
+#include "telemetry/ghst.h"
+
+#define GHST_CYCLETIME_US                   100000 		// 10x/sec
+#define GHST_FRAME_PACK_PAYLOAD_SIZE			10
+#define GHST_FRAME_LENGTH_CRC					1
+#define GHST_FRAME_LENGTH_TYPE					1
+
+static bool ghstTelemetryEnabled;
+static bool deviceInfoReplyPending;
+static uint8_t ghstFrame[GHST_FRAME_SIZE_MAX];
+
+static void ghstInitializeFrame(sbuf_t *dst)
+{
+    dst->ptr = ghstFrame;
+    dst->end = ARRAYEND(ghstFrame);
+
+    sbufWriteU8(dst, GHST_ADDR_RX);
+}
+
+static void ghstFinalize(sbuf_t *dst)
+{
+    crc8_dvb_s2_sbuf_append(dst, &ghstFrame[2]); // start at byte 2, since CRC does not include device address and frame length
+    sbufSwitchToReader(dst, ghstFrame);
+    // write the telemetry frame to the receiver.
+    ghstRxWriteTelemetryData(sbufPtr(dst), sbufBytesRemaining(dst));
+}
+
+static int ghstFinalizeBuf(sbuf_t *dst, uint8_t *frame)
+{
+    crc8_dvb_s2_sbuf_append(dst, &ghstFrame[2]); // start at byte 2, since CRC does not include device address and frame length
+    sbufSwitchToReader(dst, ghstFrame);
+    const int frameSize = sbufBytesRemaining(dst);
+    for (int ii = 0; sbufBytesRemaining(dst); ++ii) {
+        frame[ii] = sbufReadU8(dst);
+    }
+    return frameSize;
+}
+
+// Battery (Pack) status
+void ghstFramePackTelemetry(sbuf_t *dst)
+{
+    // use sbufWrite since CRC does not include frame length
+    sbufWriteU8(dst, GHST_FRAME_PACK_PAYLOAD_SIZE + GHST_FRAME_LENGTH_CRC + GHST_FRAME_LENGTH_TYPE);
+    sbufWriteU8(dst, 0x23);						// GHST_DL_PACK_STAT
+
+    if (telemetryConfig()->report_cell_voltage) {
+    	sbufWriteU16(dst, getBatteryAverageCellVoltage());		// units of 10mV
+    } else {
+    	sbufWriteU16(dst, getBatteryVoltage());
+    }
+    sbufWriteU16(dst, getAmperage());							// units of 10mA
+
+    sbufWriteU16(dst, getMAhDrawn());							// units of 10mAh
+
+    sbufWriteU8(dst, 0x00);						// Rx Voltage, units of 100mV (not passed from BF, added in Ghost Rx)
+
+    sbufWriteU8(dst, 0x00);			            // tbd1
+    sbufWriteU8(dst, 0x00);						// tbd2
+    sbufWriteU8(dst, 0x00);						// tbd3
+}
+
+#define BV(x)  (1 << (x))                       // bit value
```

Comment:
No need to define your own here - use `BIT()` from `util.h`.

Reply:
Done